### PR TITLE
feat(connector): Add connector cashtocode

### PR DESCRIPTION
**Pull request by BallaNitesh**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Add connector cashtocode of rewards payment method.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

I have manually tested the APIs through postman. 

image
image
image
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [x] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/cashtocode.rs [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+    ) -> CustomResult<String, errors::ConnectorError> {
+        Err(errors::ConnectorError::NotImplemented("get_url method".to_string()).into())
+    }
+
+    fn build_request(
+        &self,
+        _req: &types::PaymentsSyncRouterData,
+        _connectors: &settings::Connectors,
+    ) -> CustomResult<Option<services::Request>, errors::ConnectorError> {
+        Ok(None)
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cashtocode.rs:269]:
```suggestion
Err(errors::ConnectorError::FlowNotSupported {
            flow: "Capture".to_owned(),
            connector: "Zen".to_owned(),
        }
        .into())
```

Just keep build_request method rest methods can be removed, As from the core build_request() will be called

File: crates/router/src/connector/cashtocode.rs [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+        &self,
+        res: Response,
+    ) -> CustomResult<ErrorResponse, errors::ConnectorError> {
+        self.build_error_response(res)
+    }
+}
+
+impl ConnectorIntegration<api::Capture, types::PaymentsCaptureData, types::PaymentsResponseData>
+    for Cashtocode
+{
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cashtocode.rs:295]: impl "build_request" method and throw errors::ConnectorError::FlowNotSupported

File: crates/router/src/connector/cashtocode.rs [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+}
+
+impl ConnectorIntegration<api::Void, types::PaymentsCancelData, types::PaymentsResponseData>
+    for Cashtocode
+{
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cashtocode.rs:300]: impl "build_request" method and throw errors::ConnectorError::FlowNotSupported

File: crates/router/src/connector/cashtocode.rs [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+        Ok(signature)
+    }
+
+    async fn get_webhook_source_verification_merchant_secret(
+        &self,
+        db: &dyn StorageInterface,
+        merchant_id: &str,
+    ) -> CustomResult<Vec<u8>, errors::ConnectorError> {
+        let key = format!("wh_mer_sec_verification_{}_{}", self.id(), merchant_id);
+        let secret = db
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cashtocode.rs:320]:
```suggestion
let secret = match db.find_config_by_key(&key).await {
            Ok(config) => Some(config),
            Err(e) => {
                crate::logger::warn!("Unable to fetch merchant webhook secret from DB: {:#?}", e);
                None
            }
        };
```

Use `db.find_config_by_key` which will check the key in both redis and db.

File: crates/router/src/connector/cashtocode.rs [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+        Ok(api_models::webhooks::ObjectReferenceId::PaymentId(
+            api_models::payments::PaymentIdType::ConnectorTransactionId(webhook.transaction_id),
+        ))
+    }
+
+    fn get_webhook_event_type(
+        &self,
+        _request: &api::IncomingWebhookRequestDetails<'_>,
+    ) -> CustomResult<api::IncomingWebhookEvent, errors::ConnectorError> {
+        Ok(api::IncomingWebhookEvent::PaymentIntentSuccess)
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cashtocode.rs:374]: Do we receive web hooks only in success scenario?

File: crates/router/src/connector/cashtocode.rs [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+    }
+
+    fn get_webhook_resource_object(
+        &self,
+        request: &api::IncomingWebhookRequestDetails<'_>,
+    ) -> CustomResult<serde_json::Value, errors::ConnectorError> {
+        let webhook: transformers::CashtocodeIncomingWebhook = request
+            .body
+            .parse_struct("CashtocodeIncomingWebhook")
+            .change_context(errors::ConnectorError::WebhookReferenceIdNotFound)?;
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cashtocode.rs:384]:
```suggestion
            .change_context(errors::ConnectorError:: WebhookBodyDecodingFailed)?;
```

File: crates/router/src/connector/cashtocode.rs [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+        let res_json =
+            utils::Encode::<transformers::CashtocodeIncomingWebhook>::encode_to_value(&webhook)
+                .change_context(errors::ConnectorError::WebhookResourceObjectNotFound)?;
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cashtocode.rs:387]:
```suggestion
                .change_context(errors::ConnectorError:: WebhookBodyDecodingFailed)?;
```

File: crates/router/src/connector/cashtocode.rs [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+    }
+
+    fn get_webhook_api_response(
+        &self,
+        request: &api::IncomingWebhookRequestDetails<'_>,
+    ) -> CustomResult<services::api::ApplicationResponse<serde_json::Value>, errors::ConnectorError>
+    {
+        let status = "EXECUTED".to_string();
+        let id = self
+            .get_webhook_object_reference_id(request)
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cashtocode.rs:399]: Why can't we convert request  into transformers::CashtocodeObjectId and then get id from transaction_id, So that we can avoid checks you added from line 401 to 407

File: crates/router/src/connector/cashtocode/transformers.rs [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    connector::utils::RouterData,
+    core::errors,
+    services,
+    types::{self, api, storage::enums},
+};
+
+//TODO: Fill the struct with respective fields
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cashtocode/transformers.rs:12]: Keep Todo's only at required places

File: crates/router/tests/connectors/cashtocode.rs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+            }),
+            return_url: Some("https://google.com".to_owned()),
+            ..Default::default()
+        })
+    }
+}
+
+//fetch payurl for payment's create
+#[actix_web::test]
+async fn should_fetch_pay_url() {
```
Comment (ArjunKarthik) [inline crates/router/tests/connectors/cashtocode.rs:89]: Can we add unit test for Voucher also?

File: crates/router/src/connector/cashtocode/transformers.rs [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+#[derive(Default, Debug, Serialize, Eq, PartialEq)]
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cashtocode/transformers.rs:13]: Do we need Default, Eq, PartialEq?

File: crates/router/src/connector/cashtocode/transformers.rs [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+            types::ConnectorAuthType::HeaderKey { api_key } => Ok(Self {
+                api_key: api_key.to_string(),
+            }),
+            _ => Err(errors::ConnectorError::FailedToObtainAuthType.into()),
+        }
+    }
+}
+// PaymentsResponse
+//TODO: Append the remaining status flags
+#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
```
Comment (ArjunKarthik) [inline crates/router/src/connector/cashtocode/transformers.rs:105]: Please use only required tags. Verify for other structs also

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Comment by ArjunKarthik** [comment]
@BallaNitesh Can you please add the Assignees and Labels to the PR?

File: crates/router/src/connector/cashtocode/transformers.rs (hunk 11 above)
Comment (BallaNitesh) [reply crates/router/src/connector/cashtocode/transformers.rs:13]: Default might be needed for tests. Removing Eq and partialEq

File: crates/router/tests/connectors/cashtocode.rs (hunk 10 above)
Comment (BallaNitesh) [reply crates/router/tests/connectors/cashtocode.rs:89]: Sure.

File: crates/router/src/connector/cashtocode.rs (hunk 5 above)
Comment (BallaNitesh) [reply crates/router/src/connector/cashtocode.rs:374]: Yes

**Review by ArjunKarthik — DISMISSED** [review]

File: crates/router/src/connector/cashtocode/transformers.rs [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+    fn try_from(
+        item: types::ResponseRouterData<
+            F,
+            CashtocodePaymentsSyncResponse,
+            T,
+            types::PaymentsResponseData,
+        >,
+    ) -> Result<Self, Self::Error> {
+        Ok(Self {
+            status: enums::AttemptStatus::Charged,
```
Comment (jagan-jaya) [inline crates/router/src/connector/cashtocode/transformers.rs:191]: why is the Charged value hardcoded? we can have processing and failed status also in PSync right?

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/cashtocode/transformers.rs (hunk 13 above)
Comment (BallaNitesh) [reply crates/router/src/connector/cashtocode/transformers.rs:191]: This is to the handle response of the webhook. Webhook is only received when it succeeds.

File: crates/router/src/connector/cashtocode.rs [hunk 14]
```diff
@@ -0,0 +1,467 @@
...
+        Ok(api_models::webhooks::ObjectReferenceId::PaymentId(
+            api_models::payments::PaymentIdType::ConnectorTransactionId(webhook.transaction_id),
+        ))
+    }
+
+    fn get_webhook_event_type(
+        &self,
+        _request: &api::IncomingWebhookRequestDetails<'_>,
+    ) -> CustomResult<api::IncomingWebhookEvent, errors::ConnectorError> {
+        Ok(api::IncomingWebhookEvent::PaymentIntentSuccess)
```
[the remark this answers is not in the document]
Comment (BallaNitesh) [reply crates/router/src/connector/cashtocode.rs:389]: There is no failure scenario. We only have have Authentication_Pending and Succeeded. Failure is based on time delay if the txn is not succeeded within 7 days.

**Review by jagan-jaya — APPROVED** [review]

**Review by ArjunKarthik — DISMISSED** [review]

File: crates/router/src/connector/cashtocode/transformers.rs [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+    Succeeded,
+    #[default]
+    Processing,
+}
+
+impl From<CashtocodePaymentStatus> for enums::AttemptStatus {
+    fn from(item: CashtocodePaymentStatus) -> Self {
+        match item {
+            CashtocodePaymentStatus::Succeeded => Self::Charged,
+            CashtocodePaymentStatus::Processing => Self::AuthenticationPending,
```
[the remark this answers is not in the document]
Comment (BallaNitesh) [reply crates/router/src/connector/cashtocode/transformers.rs:113]: Yes.

File: crates/router/src/connector/cashtocode/transformers.rs [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+            CashtocodePaymentsSyncResponse,
+            T,
+            types::PaymentsResponseData,
+        >,
+    ) -> Result<Self, Self::Error> {
+        Ok(Self {
+            status: enums::AttemptStatus::Charged,
+            response: Ok(types::PaymentsResponseData::TransactionResponse {
+                resource_id: types::ResponseId::ConnectorTransactionId(
+                    item.data.attempt_id.clone(),
```
[the remark this answers is not in the document]
Comment (BallaNitesh) [reply crates/router/src/connector/cashtocode/transformers.rs:192]: Connector doesn't send us any transaction id. It only sends pay_url in the response.

File: loadtest/config/development.toml [hunk 17]
```diff
@@ -98,6 +99,7 @@ zen.base_url = "https://api.zen-test.com/"
 
 [connectors.supported]
 wallets = ["klarna", "braintree", "applepay"]
+rewards = ["cashtocode",]
```
[the remark this answers is not in the document]
Comment (BallaNitesh) [reply loadtest/config/development.toml:102]: Not using it anywhere.

File: crates/router/src/connector/cashtocode/transformers.rs [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+        let redirection_data = services::RedirectForm::Form {
+            endpoint: item.response.pay_url.clone(),
+            method: services::Method::Post,
+            form_fields: Default::default(),
+        };
+        Ok(Self {
+            status: enums::AttemptStatus::AuthenticationPending,
+            response: Ok(types::PaymentsResponseData::TransactionResponse {
+                resource_id: types::ResponseId::ConnectorTransactionId(
+                    item.data.attempt_id.clone(),
```
[the remark this answers is not in the document]
Comment (BallaNitesh) [reply crates/router/src/connector/cashtocode/transformers.rs:157]: This is answered here. https://github.com/juspay/hyperswitch/pull/1429/files#r1238327250

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by jagan-jaya — DISMISSED** [review]

**Review by ArjunKarthik — DISMISSED** [review]

File: crates/router/tests/connectors/connector_auth.rs [hunk 19]
```diff
@@ -48,7 +49,6 @@ pub struct ConnectorAuthentication {
 impl ConnectorAuthentication {
     #[allow(clippy::expect_used)]
     pub(crate) fn new() -> Self {
-        // Do `export CONNECTOR_AUTH_FILE_PATH="/hyperswitch/crates/router/tests/connectors/sample_auth.toml"`
```
Comment (SanchithHegde) [inline crates/router/tests/connectors/connector_auth.rs:51]: Why was this line removed?

File: crates/router/src/connector/cashtocode.rs [hunk 20]
```diff
@@ -0,0 +1,461 @@
...
+    match payment_method_type
+        .clone()
+        .ok_or_else(conn_utils::missing_field_err("payment_method_type"))
+    {
+        Ok(reward_type) => match reward_type {
+            storage::enums::PaymentMethodType::ClassicReward => match auth_type {
+                types::ConnectorAuthType::BodyKey { api_key, key1: _ } => Ok(vec![(
+                    headers::AUTHORIZATION.to_string(),
+                    format!("Basic {}", api_key).into_masked(),
+                )]),
```
Comment (SanchithHegde) [inline crates/router/src/connector/cashtocode.rs:57]:
```suggestion
                types::ConnectorAuthType::BodyKey { api_key, .. } => Ok(vec![(
                    headers::AUTHORIZATION.to_string(),
                    format!("Basic {api_key}").into_masked(),
                )]),
```

File: crates/router/src/connector/cashtocode.rs [hunk 21]
```diff
@@ -0,0 +1,461 @@
...
+                _ => Err(errors::ConnectorError::FailedToObtainAuthType.into()),
+            },
+            storage::enums::PaymentMethodType::Evoucher => match auth_type {
+                types::ConnectorAuthType::BodyKey { api_key: _, key1 } => Ok(vec![(
+                    headers::AUTHORIZATION.to_string(),
+                    format!("Basic {}", key1.to_owned()).into_masked(),
+                )]),
```
Comment (SanchithHegde) [inline crates/router/src/connector/cashtocode.rs:64]:
```suggestion
                types::ConnectorAuthType::BodyKey { key1, .. } => Ok(vec![(
                    headers::AUTHORIZATION.to_string(),
                    format!("Basic {key1}").into_masked(),
                )]),
```

You shouldn't need the clone.

File: crates/api_models/src/payments.rs [hunk 22]
```diff
@@ -940,6 +943,13 @@ pub struct CardResponse {
     exp_year: String,
 }
 
+#[derive(Debug, Clone, Eq, PartialEq, serde::Serialize, serde::Deserialize, ToSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct RewardData {
+    /// The merchant ID with which we have to call the connector
+    pub mid: String,
```
[the remark this answers is not in the document]
Comment (BallaNitesh) [reply crates/api_models/src/payments.rs:950]: Sure

File: crates/router/src/core/payments/flows.rs [hunk 23]
```diff
@@ -137,6 +137,7 @@ impl<const T: u8>
 }
 
 default_imp_for_complete_authorize!(
+    connector::Cashtocode,
```
[the remark this answers is not in the document]
Comment (BallaNitesh) [reply crates/router/src/core/payments/flows.rs:140]: Sure

File: crates/router/tests/connectors/connector_auth.rs (hunk 19 above)
Comment (BallaNitesh) [reply crates/router/tests/connectors/connector_auth.rs:51]: Was asked to remove in one of the Peer review.

**Review by SanchithHegde — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]
